File: ABC/B/258/B.py
n = int(input())
l = list()
vector = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]]
max_val = -1
for i in range(n):
    s = input()
    l.append(s)
for i in range(n):
    for j in range(n):
        for direct in vector:
            x, y = i, j
            num = str()
            for k in range(n):
                x += direct[0]
                y += direct[1]
                num += l[x%n][y%n]
            num = int(num)
            max_val = max(max_val, num)
print(max_val)

File: ABC/B/258/B.rb
direct = [-1, 0, 1].repeated_permutation(2).to_a
direct.delete([0, 0])

n = gets.chomp.to_i
array = Array.new
n.times do
    array << gets.chomp
end

max_val = -1
for i in 0...n do
    for j in 0...n do
        direct.each do |vec|
            s = [i, j].dup
            str = ""
            n.times do
                i += vec[0]
                j += vec[1]
                str += array[i%n][j%n]
            end
            max_val = [max_val, str.to_i].max
        end
    end
end
p max_val